beps/0007-cli-yaml-validation/README.md

```diff
@@ -0,0 +1,90 @@
+---
+title: CLI YAML Validation
+status: provisional
+authors:
+  - '@drodil'
+project-areas:
+  - cli
+creation-date: 2024-03-23
+---
+
+# BEP: CLI YAML Validation
+
+[**Discussion Issue**](https://github.com/backstage/backstage/issues/23817)
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+- [Release Plan](#release-plan)
+- [Dependencies](#dependencies)
+- [Alternatives](#alternatives)
+
+## Summary
+
+This BEP proposes to add a new command to the Backstage CLI that will validate Backstage YAML files in a repository. This command will be useful for contributors to ensure that their changes are valid before submitting a pull request.
+
+## Motivation
+
+Currently, there is no way to validate the YAML files in the repositories containing Backstage entities. This can lead to errors that are only discovered after the changes have been merged and the entity has been imported to the Backstage instance by an entity provider. By adding a validation command to the CLI, contributors can ensure that their changes are valid before submitting a pull request.
+
+Additionally, the validation command can be used in CI/CD pipelines to ensure that the YAML files are valid before merging changes to the main branch.
+
+The main motivation for this BEP is to improve the developer experience and reduce the number of errors that occur when importing entities to the Backstage instance.
+
+### Goals
+
+- Add a new command to the Backstage CLI that can be used to validate the YAML files in the repository.
```

Comment:
Modularizing the backstage CLI is a good idea. Another option is to split it up in separate commands but it might end up having too much same code shared between different tools. All of the things mentioned; catalog validation, scaffolder dry-runs, and techdocs-cli, are necessary for people who want to contribute to a Backstage app. 

I would even go further like having the possibility to check entities from the command line using CLI or doing some admin tasks like triggering tasks. But I think all this requires first a clear direction of where the CLI tooling is going.

Reply:
Main concern is actually all of the different dependencies that have to been installed. Especially the build deps but also things like Jest and ESLint are pretty large and takes a long time to install. Modularizing into separate packages means you only need to install the dependencies that are needed for the particular commands you want to have available. It also opens up for 3rd party extensions of the CLI.

+1 to the admin tasks too. There's a bit of a risk that even with a modularized CLI there might be confusion when we cover the full spectrum use-cases with a single entry-point, but I think that's something we can think about if we chose that path. That's mostly in naming and promotion imo

Reply:
So going forward, should we start with this BEP anyways and change it on the way? Maybe you need some internal discussion with the core team about the future of the CLI first?

Reply:
I think the BEP that we need right now for it to be actionable is one around modularization of the CLI. The content in here I think makes sense to move to a feature request issue though, to be woven into that work in the future?

I don't think we want to build this feature into any of our existing CLIs and creating a new one isn't where I think we should be going either. That effectively means that we can't build this right now until we've done some more foundational work.